add a checker for misleading unicode (#5311)
PyCQA/pylint: doc/whatsnew/2.13.rst

New checkers
============

* Added several checkers to deal with unicode security issues
  (see `Trojan Sources <https://trojansource.codes/>`_ and
  `PEP 672 <https://www.python.org/dev/peps/pep-0672/>`_ for details) that also
  concern the readability of the code. In detail the following checks were added:

>   * ``bad-file-encoding`` checks that the file is encoded in UTF-8 as suggest by

Reviewer: ```suggestion
  * ``bad-file-encoding`` checks that the file is encoded in UTF-8 as suggested by
```

    `PEP8 <https://www.python.org/dev/peps/pep-0008/#id20>`_.
    UTF-16 and UTF-32 are `not supported by Python <https://bugs.python.org/issue1503789>`_
    at the moment. If this ever changes
    ``invalid-unicode-codec`` checks that they aren't used, to allow for backwards
    compatibility.

  * ``bidirectional-unicode`` checks for bidirectional unicode characters that
    could make code execution different than what the user expects.

  * ``invalid-character-backspace``, ``invalid-character-carriage-return``,
    ``invalid-character-sub``, ``invalid-character-esc``,
    ``invalid-character-zero-width-space`` and ``invalid-character-nul``

>     to check for possibly harmful not escaped characters.

Reviewer: ```suggestion
    to check for possibly harmful unescaped characters.
```

> Closes #5281

Reviewer: ```suggestion
  Closes #5281
```

New checkers
============

* Added several checkers to deal with unicode security issues
  (see `Trojan Sources <https://trojansource.codes/>`_ and
  `PEP 672 <https://www.python.org/dev/peps/pep-0672/>`_ for details) that also
  concern the readability of the code. In detail the following checks were added:

  * ``invalid-unicode-codec`` checks that UTF-16 and UTF-32 is not used to encode the
    file. These encodings are currently not valid for Python but could be in the future.
    As this could break the other checks, we want to make users aware of the issue.

  * ``bidirectional-unicode`` checks for bidirectional unicode characters that could
    obstruct real code execution.

  * ``invalid-ascii-backspace``, ``invalid-ascii-carriage-return``, ``invalid-ascii-sub``,
    ``invalid-ascii-esc`` and ``invalid-ascii-nul`` to check for possible unwanted
    ASCII characters.

 Closes #5281

* ``non-ascii-name`` check was rewritten to ensure really __all__ Python names are ASCII as

>    as non ASCII characters could be homoglyphs (look a like characters) and hard to

Reviewer 2: ```suggestion
   as non ASCII characters could be homoglyphs (look alike characters) and hard to
```

  `PEP 672 <https://www.python.org/dev/peps/pep-0672/>`_ for details) that also
  concern the readability of the code. In detail the following checks were added:

  * ``bad-file-encoding`` checks that the file is encoded in UTF-8 as suggest by
    `PEP8 <https://www.python.org/dev/peps/pep-0008/#id20>`_.
    UTF-16 and UTF-32 are `not supported by Python <https://bugs.python.org/issue1503789>`_
    at the moment. If this ever changes
    ``invalid-unicode-codec`` checks that they aren't used, to allow for backwards
    compatibility.

  * ``bidirectional-unicode`` checks for bidirectional unicode characters that
    could make code execution different than what the user expects.

  * ``invalid-character-backspace``, ``invalid-character-carriage-return``,
    ``invalid-character-sub``, ``invalid-character-esc``,
    ``invalid-character-zero-width-space`` and ``invalid-character-nul``
    to check for possibly harmful not escaped characters.

Closes #5281

* ``non-ascii-name`` check was rewritten to ensure really __all__ Python names are ASCII as
   as non ASCII characters could be homoglyphs (look alike characters) and hard to
   enter on a non specialized keyboard.

> 

Reviewer: ```suggestion
```
